include/PetriEngine/Colored/VariableVisitor.h

```diff
@@ -289,6 +295,32 @@ namespace PetriEngine {
 
                 get_variables(e, variables, varPositions, varModifierMap, tuples, false);
             }
+
+        private:
+            bool is_last_variable_expr(const std::shared_ptr<const ColorExpression> expr) const {
+                std::stack<std::shared_ptr<const ColorExpression>> stack;
+                stack.push(expr);
+
+                while (!stack.empty()) {
+                    auto currentExpr = stack.top();
```

Comment:
This will create another copy of the smart pointer. If the stack is just regular pointers, it is virtually free.